File: go-backend-interview-prep/02_data_structures/06_structs_intro/main.py
# Python equivalent of Structs (compare with main.go)
# Go struct -> Python dataclass (or simple class)

from dataclasses import dataclass


@dataclass
class User:
    name: str = ""
    email: str = ""
    age: int = 0


def new_user(name: str, email: str, age: int) -> User:
    """Factory function (same convention as Go's NewUser)."""
    return User(name=name, email=email, age=age)


def try_to_modify(u: User) -> None:
    # Dataclasses are mutable by default -- this WILL modify the original
    # (unlike Go structs which are value types and get copied)
    u.name = "CHANGED"


def main():
    # --- Create with field names ---
    u1 = User(name="Alice", email="alice@example.com", age=30)
    print("u1:", u1)

    # --- Zero-value equivalent ---
    u2 = User()
    print(f"u2 (default): {u2}")

    # --- Partial initialization ---
    u3 = User(name="Bob")
    print(f"u3 (partial): {u3}")

    # --- Factory function ---
    u4 = new_user("Charlie", "charlie@example.com", 25)
    print("u4:", u4)

    # --- Access and modify ---
    u1.age = 31
    print("u1 after birthday:", u1.age)

    # --- Python dataclasses are REFERENCE types (unlike Go structs!) ---
    print("\n--- Reference semantics ---")
    original = User(name="Dave", age=40)
    copied = original            # same object!
    copied.name = "Modified"
    print("original:", original.name)  # "Modified" -- CHANGED (unlike Go)

    # --- To copy, use dataclasses.replace or copy ---
    from copy import copy
    original2 = User(name="Dave", age=40)
    copied2 = copy(original2)
    copied2.name = "Modified"
    print("original2:", original2.name)  # "Dave" -- unchanged
    print("copied2:", copied2.name)

    # --- Passing to function mutates original ---
    print("\n--- Pass by reference ---")
    test_user = User(name="Eve")
    try_to_modify(test_user)
    print("after function:", test_user.name)  # "CHANGED" (unlike Go)

    # --- Comparison ---
    print("\n--- Comparison ---")
    a = User(name="Eve", age=28)
    b = User(name="Eve", age=28)
    print("a == b:", a == b)  # True (dataclass __eq__)


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/02_data_structures/06_structs_intro/main.go
package main

import "fmt"

// --- Define a struct type ---
type User struct {
	Name  string
	Email string
	Age   int
}

// --- Factory function (Go convention for "constructors") ---
func NewUser(name, email string, age int) User {
	return User{
		Name:  name,
		Email: email,
		Age:   age,
	}
}

func main() {
	// --- Create with field names ---
	u1 := User{Name: "Alice", Email: "alice@example.com", Age: 30}
	fmt.Println("u1:", u1)

	// --- Zero-value struct ---
	var u2 User
	fmt.Printf("u2 (zero): %+v\n", u2) // %+v shows field names

	// --- Partial initialization ---
	u3 := User{Name: "Bob"}
	fmt.Printf("u3 (partial): %+v\n", u3)

	// --- Factory function ---
	u4 := NewUser("Charlie", "charlie@example.com", 25)
	fmt.Println("u4:", u4)

	// --- Access and modify fields ---
	u1.Age = 31
	fmt.Println("u1 after birthday:", u1.Age)

	// --- Structs are VALUE types ---
	fmt.Println("\n--- Value semantics ---")
	original := User{Name: "Dave", Age: 40}
	copied := original
	copied.Name = "Modified"
	fmt.Println("original:", original.Name) // "Dave" -- unchanged
	fmt.Println("copied:", copied.Name)     // "Modified"

	// --- Passing to a function copies ---
	fmt.Println("\n--- Pass by value ---")
	tryToModify(original)
	fmt.Println("after function:", original.Name) // still "Dave"

	// --- Anonymous struct (one-off use) ---
	fmt.Println("\n--- Anonymous struct ---")
	point := struct {
		X, Y int
	}{X: 10, Y: 20}
	fmt.Println("point:", point)

	// --- Struct comparison ---
	fmt.Println("\n--- Comparison ---")
	a := User{Name: "Eve", Age: 28}
	b := User{Name: "Eve", Age: 28}
	fmt.Println("a == b:", a == b) // true (all fields are comparable)

	// --- Printing with format verbs ---
	fmt.Println("\n--- Format verbs ---")
	fmt.Printf("%%v:  %v\n", u1)
	fmt.Printf("%%+v: %+v\n", u1) // with field names
	fmt.Printf("%%#v: %#v\n", u1) // Go syntax representation
}

func tryToModify(u User) {
	u.Name = "CHANGED" // modifies the copy, not the original
}
